src/Covenant/Internal/ASGBuilder.hs

```diff
@@ -53,154 +49,193 @@ makeFieldLabelsNoPrefix ''ASGBuilderState
 -- enables hash consing.
 --
 -- @since 1.0.0
-newtype ASGBuilder (a :: Type) = ASGBuilder (State ASGBuilderState a)
+newtype ASGBuilder (a :: Type) = ASGBuilder (ExceptT ASGBuilderError (State ASGBuilderState) a)
   deriving
     ( -- | @since 1.0.0
       Functor,
       -- | @since 1.0.0
       Applicative,
       -- | @since 1.0.0
-      Monad
+      Monad,
+      -- | @since 1.0.0
+      MonadError ASGBuilderError
+    )
+    via (ExceptT ASGBuilderError (State ASGBuilderState))
+
+-- | The errors that can occur during the construction of an ASG
+--
+-- @since 1.0.0
+newtype ASGBuilderError = ATypeError TypeError
+  deriving
+    ( -- | @since 1.0.0
+      Eq,
+      -- | @since 1.0.0
+      Show
+    )
+    via TypeError
+
+-- | The errors that come up while resolving types.
+--
+-- @since 1.0.0
+data TypeError
+  = TyErrAppNotALambda ASGType
+  | TyErrAppArgMismatch ASGType ASGType -- expected, received
+  | TyErrPrimNotAConstant ASGType
+  | TyErrPrimArgMismatch (Vector TyConstant) (Vector TyConstant) -- exptected, received
+  | TyErrNonHomogenousList
```

Comment:
These constructors should have Haddocks of their own explaining their meaning. While in some cases it's fairly clear (`TyErrAppNotALambda` for example), in others, it's not `TyErrAppArgMismatch` would require me to read the source code to understand which field was which).